### DISCUSSION TITLE
Pygls

### DISCUSSION THREAD

**Discussion by bagel897** [python-lsp/python-lsp-server#206] [Ideas]
[Pygls](https://pygls.readthedocs.io/en/latest/index.html) has several advantages over the existing jsonrpc implementation
1. It uses pydantic objects instead of dicts, making future development easier. It also is type hinted.
2. It implements more of the LSP protocol, such as notification/progress support
3. It runs everything in asyncio with thread management, and has websocket support.
4. It would reduce maintenance burden on pylsp by using a common library to implement basic features.
I'm going to try to implement this sometime, but I was wondering what the opinion would be on this.

Other servers, like [anakin-language-server](https://github.com/muffinmad/anakin-language-server) and jedi language server use pygls but pylsp has the largest featureset and community.

**Comment by ccordoba12** [comment]
It looks interesting, but perhaps we could only use [JsonRPCProtocol](https://github.com/openlawlibrary/pygls/blob/master/pygls/protocol.py#L167) from it to replace our own implementation. I say it because we already provide something similar to [LanguageServerProtocol](https://github.com/openlawlibrary/pygls/blob/master/pygls/protocol.py#L549) and a rewrite would be too costly.
